contributors/design-proposals/apps/configmap-secret-gc.md

```diff
@@ -0,0 +1,271 @@
+  // gone, even if this ConfigMap is still referenced by other API resources
+  // (such as Jobs).
+  // This can only be specified at creation time.
+  // Default to false.
+  ControllersAsOwners *bool
+}
+```
+
+ConfigMap and Secret garbage collection are disabled by default for backward
+compatibility. 
+
+
+## Controllers
+
+This section is presented as a generalization of how an arbitrary controller can
+use this new API field to garbage collect the ConfigMaps/Secrets it references. 
+
+A controller can have the ConfigMaps/Secrets it references garbage collected by
+declaring ownership of those ConfigMaps/Secrets, i.e. set the ConfigMaps/Secrets
+(`.metadata.ownerReferences`).
+The [garbage collector](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/)
+will then delete the ConfigMaps/Secrets when the owners are all gone. Note that
+garbage collector only deletes objects that once have owners but no longer do.
+ConfigMaps/Secrets won't be garbage collected if they never have any owner. 
+
+Most workload controllers support rolling update, revision history and rollback
+features. To make sure rollback still works, for a controller with revision
+history feature, the ConfigMaps/Secrets ownership should belong to the
+controller's history instead of the controller itself. For example, a Deployment
+should not become the owner of any referenced ConfigMaps/Secrets, but the
+Deployment's history objects (ReplicaSets) will own those ConfigMaps/Secrets.
+
+If a controller does not support rolling update, revision history, or rollback
+(such as ReplicaSets and ReplicationControllers), it should mark itself as the
+owner of the ConfigMaps/Secrets it references in order to have them garbage
+collected when the controller is gone. 
+
+
+### Core Controllers that Support ConfigMaps/Secrets Garbage Collection
+
+*   Controllers with history: Deployments, StatefulSets, DaemonSets
+*   Controllers without history: ReplicaSets, ReplicationControllers 
+
+
+### Update ConfigMap/Secret's Owner List
+
+There are 2 possible scenarios when a referenced ConfigMap/Secret's owner list
+(`.metadata.ownerReferences`) needs to be updated:
+
+1.  After a controller API resource is created.
+1.  After a controller API resource's pod template is updated, and: 
+    1.  The controller's current history object is created, or 
+    1.  The ConfigMap/Secret references in the pod template are changed. 
+
+Note that Deployment is a special case here. Because a Deployment's revision
+history (ReplicaSets) can declare ConfigMaps/Secrets ownership directly, we
+don't need to implement it in Deployment controller code. Deployment controller
+delegates the work to ReplicaSet controller. 
+
+#### On Controller API Resource Creation
+
+When a controller API object (with or without history) exists, the controller
+will do the following:
+
+1.  Find all referenced ConfigMaps/Secrets of the following:
```

Comment:
Somewhat of a red flag in that it gives those controllers access to all secrets in the system.  

It would be better to have a controller that copies config maps and then makes mutations to controllers, without controllers being aware of it (for that particular problem)

Reply:
I like the idea to keep CM/Secret mutation outside the workload controllers if possible. It would be nice, for example, if we could have a controller that does nothing but watch ControllerRevision objects and add ownerRefs to the CMs/Secrets that they reference (if they are opted in to CM GC). However, since ControllerRevision currently only contains opaque data, we would need to modify each controller (DS/StS) to put in a list of CMs/Secrets that each ControllerRevision references.

We're also planning to consult sig-auth to see how concerned they are about such permissions, given that these controllers can already create Pods, which transitively gives them access to (at least read) all Secrets.

Reply:
If it's not allowed to give controllers access to secrets in the system, we probably cannot use ownerRefs for CM/S GC anymore.

Reply:
From the [docs](https://kubernetes.io/docs/concepts/configuration/secret/#risks), if you can create a Pod that uses the secret you can access the secret. This seems transitive through  controllers today. If I can create a Deployment that creates Pods that can access the secret I can access the secret. There is nothing to stop a user from creating a Deployment that reads a secret and serves it via http. Giving a controller the ability to mutate the secret adds an additional risk to integrity (of the secret itself not the data its protecting), but you could (in some cases) address this by creating a secret per Deployment. We probably don't want to give controllers delete permissions. That defeats the purpose of the GC, and is an unnecessary risk to availability.

Reply:
We deliberately don't enforce hard referential integrity.

kubectl currently doesn't create/update resources in any particular order, and controllers may observe changes out of order even if it did. 

However, I would like controllers, such as Deployment and ReplicaSet, to behave gracefully in the absence of dependencies, such as ConfigMap and Secret. For example, possibly they shouldn't create lots of pods that can't function.

If a resource can't properly function due to a nonexistent dependent resource, I recommend reporting that as the reason the resource isn't fully functional in status. 

This would require at least being able to confirm the resource's existence or lack thereof.

Reply:
If we're concerned about permissions, we could create a subresource for updating ownerReferences, finalizers, and possibly other metadata in order to facilitate lifecycle management without privilege escalation.

Reply:
That sounds interesting... filed #54269.